File: csharp-ads/src/ADS.DataStructures/DoublyLinkedList/DoublyLinkedList.cs
using System.Collections.Generic;
using System.Linq;

namespace ADS.DataStructures.DoublyLinkedList
{
    public sealed class DoublyLinkedList<T> where T : class
    {
        public long Count { get; private set; }

        public Node<T> Head { get; private set; }

        public Node<T> Tail { get; private set; }

        public bool Any()
        {
            return Head != null;
        }

        public void AddFront(T value)
        {
            var node = new Node<T>(value);
            node.Next = Head;

            if (Head == null)
            {
                Tail = node;
            }
            else
            {
                Head.Previous = node;
            }

            Head = node;
            Count++;
        }

        public void AddEnd(T value)
        {
            var node = new Node<T>(value);

            if (Tail == null)
            {
                Head = node;
            }
            else
            {
                Tail.Next = node;
                node.Previous = Tail;
            }

            Tail = node;
            Count++;
        }

        public Node<T> RemoveFront()
        {
            if (Head == null) return null;

            var removedNode = Head;

            if (Head.Next == null)
                Tail = null;
            else
                Head.Next.Previous = null;

            Head = Head.Next;
            removedNode.Next = null;
            Count--;

            return removedNode;
        }

        public Node<T> RemoveEnd()
        {
            if (Head == null) return null;

            var removedNode = Tail;

            if (Tail.Previous == null)
                Head = null;
            else
                Tail.Previous.Next = null;

            Tail = Tail.Previous;
            removedNode.Previous = null;
            Count--;

            return removedNode;
        }

        public IEnumerable<T> ToEnumerable()
        {
            var node = Head;

            while (node != null)
            {
                yield return node.Value;
                node = node.Next;
            }
        }

        public List<T> ToList()
        {
            return ToEnumerable().ToList();
        }

        public override string ToString()
        {
            var nodes = string.Join(" <-> ", ToList());
            var className = nameof(DoublyLinkedList<T>);

            return $"{className} [ {nodes} ]";
        }
    }
}

File: javascript-ads/src/data-structures/doubly-linked-list/DoublyLinkedList.js
const { Node } = require('./Node');

class DoublyLinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
        this.count = 0;
    }

    any() {
        return this.head !== null;
    }

    addFront(value) {
        const node = new Node(value);
        node.next = this.head;

        if (this.head === null) {
            this.tail = node;
        } else {
            this.head.previous = node;
        }

        this.head = node;
        this.count++;
    }

    addEnd(value) {
        const node = new Node(value);

        if (this.tail === null) {
            this.head = node;
        } else {
            this.tail.next = node;
            node.previous = this.tail;
        }

        this.tail = node;
        this.count++;
    }

    removeFront() {
        if (this.head === null) {
            return null;
        }

        const deletedNode = this.head;

        if (this.head.next == null) {
            this.tail = null;
        } else {
            this.head.next.previous = null;
        }

        this.head = this.head.next;
        deletedNode.next = null;
        this.count--;

        return deletedNode;
    }

    removeEnd() {
        if (this.tail === null) {
            return null;
        }

        const deletedNode = this.tail;

        if (this.tail.previous === null) {
            this.head = null;
        } else {
            this.tail.previous.next = null;
        }

        this.tail = this.tail.previous;
        deletedNode.previous = null;
        this.count--;

        return deletedNode;
    }

    toArray() {
        const nodes = [];

        let currentNode = this.head;

        while (currentNode != null) {
            nodes.push(currentNode.value);
            currentNode = currentNode.next;
        }

        return nodes;
    }

    toString() {
        const nodes = this.toArray();
        return `DoublyLinkedList [ ${nodes.join(' <-> ')} ]`;
    }
}

module.exports = { DoublyLinkedList };
